app/uk/gov/hmrc/tai/config/CustomErrorHandler.scala

```diff
@@ -0,0 +1,225 @@
+          "NOT_FOUND",
+          s"The resource `${request.uri}` has not been found",
+          reportAs = NOT_FOUND
+        ).toResult
+
+      case BAD_REQUEST =>
+        auditConnector.sendEvent(
+          httpAuditEvent.dataEvent(
+            eventType = "ServerValidationError",
+            transactionName = "Request bad format exception",
+            request = request,
+            detail = Map.empty
+          )
+        )
+
+        val msg =
+          if (suppress4xxErrorMessages) "Bad request"
+          else constructErrorMessage(message)
+
+        ApiResponse(
+          "BAD_REQUEST",
+          msg + s" [auditSource=${appConfig.appName}, X-Request-ID=$requestId]",
+          reportAs = BAD_REQUEST
+        ).toResult
+
+      case _ =>
+        auditConnector.sendEvent(
+          httpAuditEvent.dataEvent(
+            eventType = "ClientError",
+            transactionName = s"A client error occurred, status: $statusCode",
+            request = request,
+            detail = Map.empty
+          )
+        )
+
+        Status(statusCode)(
+          ApiResponse(
+            "CLIENT_ERROR",
+            s"Other error [auditSource=${appConfig.appName}, X-Request-ID=$requestId]",
+            reportAs = statusCode
+          )
+        )
+    }
+    Future.successful(result)
+  }
+
+  override def onServerError(request: RequestHeader, ex: Throwable): Future[Result] = {
+    implicit val headerCarrier: HeaderCarrier = hc(request)
+    val requestId = headerCarrier.requestId.map(_.value).getOrElse("None")
+    val eventType = ex match {
+      case _: NotFoundException      => "ResourceNotFound"
+      case _: AuthorisationException => "ClientError"
+      case _: JsValidationException  => "ServerValidationError"
+      case _                         => "ServerInternalError"
+    }
+
+    logger.error(
+      s"! Internal server error, for (${request.method}) [auditSource=${appConfig.appName}, X-Request-ID=$requestId -> ",
+      ex
+    )
+
+    auditConnector.sendEvent(
+      httpAuditEvent.dataEvent(
+        eventType = eventType,
+        transactionName = "Unexpected error",
+        request = request,
+        detail = Map("transactionFailureReason" -> ex.getMessage)
+      )
+    )
+
+    Future.successful(
+      ApiResponse(
+        "INTERNAL_ERROR",
+        s"An error has occurred. This has been audited with auditSource=${appConfig.appName}, X-Request-ID=$requestId",
+        reportAs = INTERNAL_SERVER_ERROR
+      ).toResult
+    )
+  }
+}
+
+object CustomErrorHandler {
+  private case class ErrorView(url: String, statusCode: Int)
+
+  private object ErrorView {
+    implicit val format: OFormat[ErrorView] = Json.format[ErrorView]
+  }
+  private case class ApiResponse(
+    code: String,
+    message: String,
+    errorView: Option[ErrorView] = None,
+    redirect: Option[String] = None,
+    reportAs: Int
+  ) {
+    def toResult: Result = Status(reportAs)(this)
+  }
+
+  private object ApiResponse {
```

Comment:
this is from pertax. Tai probably has an other structure for its json response.
It's here: https://github.com/hmrc/tai/blob/04d509b9efea4ee148b8962f7ebda8c176f0fd13/app/uk/gov/hmrc/tai/model/api/ApiResponse.scala#L21

However it is not suitable for an error response. I believe at the moment the error body is raw text, not json?

We could keep this internal object but rename it and just use the message. The rest has no use here.